Fix indexing IdentityUnitRange with Integers and Boolean ranges

jishnub opened the pull request:
Fixes #39997 

Now:
```julia
julia> r = Base.IdentityUnitRange(2:1000)
Base.IdentityUnitRange(2:1000)

julia> r[Int32(5)]
5

julia> r[Int64(5)]
5
```

jishnub committed:
Fix indexing IdentityUnitRange with Integers

jishnub committed:
LogicalIndexing for an IdentityUnitRange

jishnub commented:
I've also added logical indexing for an `IdentityUnitRange` so that now its behavior matches that of an `UnitRange`.

```julia
julia> Base.IdentityUnitRange(1:2)[false:true]
2:2

julia> (1:2)[false:true]
2:2
```

mbauman commented:
Oh dang, here I was thinking that the default `getindex(::AbstractArray, ::Integer)` definition would convert everything to `Int`s for us; I had forgotten about the default `getindex(::AbstractRange, ::Integer)`.

mbauman commented on base/indices.jl:
```diff
@@ -394,9 +394,52 @@ last(S::IdentityUnitRange) = last(S.indices)
-getindex(S::IdentityUnitRange, i::StepRange{<:Integer}) = (@_inline_meta; @boundscheck checkbounds(S, i); i)
+function getindex(S::IdentityUnitRange, i::Integer)
+    @_inline_meta
+    i isa Bool && throw(ArgumentError("invalid index: $i of type Bool"))
```
```suggestion
    j = to_indices(r, (i,))[1]
```

mbauman commented on base/indices.jl:
```diff
@@ -394,9 +394,52 @@ last(S::IdentityUnitRange) = last(S.indices)
+                return range(first(S), length=0)
+            end
+        else # length(i) == 2
+            return range(first(S) + one(first(S)), length=1)
```
```suggestion
            return range(last(S), length=1)
```

mbauman commented on base/indices.jl:
```diff
@@ -394,9 +394,52 @@ last(S::IdentityUnitRange) = last(S.indices)
+            return range(first(S), length = 0)
+        elseif length(i) == 1
+            if first(i)
+                return range(first(S), length = length(S))
```
```suggestion
                return range(first(S), length = 1)
```

mbauman commented on base/indices.jl:
```diff
@@ -394,9 +394,52 @@ last(S::IdentityUnitRange) = last(S.indices)
+            return range(first(S) + one(first(S)), length=1)
+        end
+    else
+        return i
```
Not sure about this one — but perhaps it should be:

```suggestion
        return Int.(i)
```

jishnub replied:
Thanks for the suggestions! Perhaps the correct choice here is `map(Int, i)` as it has special methods defined to return ranges, whereas broadcasting currently converts a range to a `Vector`.

```julia
julia> Int.(big(2):big(3))
2-element Vector{Int64}:
 2
 3

julia> map(Int, big(2):big(3))
2:3
```

I'll create a new commit with all the changes incorporated.

mbauman commented on base/indices.jl:
```diff
@@ -394,9 +394,52 @@ last(S::IdentityUnitRange) = last(S.indices)
+    @boundscheck checkbounds(S, i)
+    if eltype(i) === Bool
+        # logical indexing
+        if length(i) == 0
+            return range(first(S), step=one(eltype(S)), length=0)
+        elseif length(i) == 1
+            if first(i)
+                return range(first(S), step=one(eltype(S)), length=1)
+            else
+                return range(first(S), step=one(eltype(S)), length=0)
+            end
+        else # length(i) == 2
+            return range(first(S) + one(first(S)), step=one(eltype(S)), length=1)
+        end
```
```suggestion
        if length(i) == 0
            return range(first(S), step=oneunit(eltype(S)), length=0)
        elseif length(i) == 1
            if first(i)
                return range(first(S), step=oneunit(eltype(S)), length=1)
            else
                return range(first(S), step=oneunit(eltype(S)), length=0)
            end
        else # length(i) == 2
            return range(last(S), step=oneunit(eltype(S)), length=1)
        end
```

vtjnash commented on base/indices.jl:
```diff
@@ -394,9 +394,52 @@ last(S::IdentityUnitRange) = last(S.indices)
+    @_inline_meta
+    i isa Bool && throw(ArgumentError("invalid index: $i of type Bool"))
+    @boundscheck checkbounds(S, i)
+    i
```
```suggestion
    return i % Int
```

Refs: https://github.com/JuliaLang/julia/issues/40013

mbauman commented on base/indices.jl:
```diff
@@ -394,9 +394,52 @@ last(S::IdentityUnitRange) = last(S.indices)
+            return range(first(S) + one(first(S)), step=one(eltype(S)), length=1)
+        end
+    else
+        return i
```
Similar to above:

```suggestion
        return Int.(i)
```

jishnub commented on base/indices.jl:
```diff
@@ -394,9 +394,52 @@ last(S::IdentityUnitRange) = last(S.indices)
+    @_inline_meta
+    i isa Bool && throw(ArgumentError("invalid index: $i of type Bool"))
+    @boundscheck checkbounds(S, i)
+    i
```
This might break indexing if the `IdentityUnitRange` wraps a `BigInt` `UnitRange` though. 

```julia
julia> IdentityUnitRange(1:big(2)^65)[end]
ERROR: InexactError: Int64(36893488147419103232)
Stacktrace:
 [1] Type
   @ ./gmp.jl:362 [inlined]
 [2] convert
   @ ./number.jl:7 [inlined]
 [3] to_index
   @ ./indices.jl:292 [inlined]
 [4] to_index
   @ ./indices.jl:277 [inlined]
 [5] to_indices
   @ ./indices.jl:333 [inlined]
 [6] to_indices
   @ ./indices.jl:330 [inlined]
 [7] getindex(S::IdentityUnitRange{UnitRange{BigInt}}, i::BigInt)
   @ Base ~/julia/base/indices.jl:400
 [8] top-level scope
   @ REPL[89]:1
```

jishnub replied:
Sorry perhaps I've replied to the wrong comment, I meant using `to_indices`

mbauman replied:
Yeah, but we declare the `IdentityUnitRange` as a `<: AbstractVector{Int}`.  Perhaps we should prevent construction of non-Int ranges or dynamically grab the eltype from the wrapped range.

jishnub replied:
Does it make sense to define `IdentityUnitRange` as `IdentityUnitRange{T <:Integer, I <:AbstractUnitRange{T}} <: AbstractUnitRange{T}` instead? This way it'll work correctly for subtypes of `Integer`, and indexing may coerce to the type `T`. This is similar to how `IdOffsetRange` is implemented in `OffsetArrays`.

mbauman replied:
Yes, that'd be best (and is precisely what I was meaning with that latter option).

jishnub replied:
Although this appears to be breaking the precompile directives of `OffsetArrays` presently

jishnub committed:
use to_indices and oneunit

jishnub committed:
add eltype as a paramter in IdentityUnitRange

jishnub commented:
After 2042090:

```julia
julia> r = IdentityUnitRange(1:big(2)^65)
Base.IdentityUnitRange(1:36893488147419103232)

julia> r[1] |> eltype
BigInt

julia> first(r) |> eltype
BigInt
```

so the types become consistent. This is somewhat breaking though if packages have used `IdentityUnitRange{<:AbstractUnitRange}` specifically, although a quick search on juliahub seems to indicate that it's [primarily `OffsetArrays` that had uses this](https://juliahub.com/ui/RepoSearch?q=IdentityUnitRange%7B&r=true), although there are some others.

jishnub committed:
Fix references to IdentityUnitRange in OffsetArrays

jishnub committed:
non-breaking definition of IdentityUnitRange

Add a type parameter to IdentityUnitRange that acts as the eltype.
This resolves issues such as indexing and iteration producing elements
having a type that differs from eltype of the range.

jishnub commented:
I've flipped the order of the type parameters, so this should not be breaking anymore. I've checked locally that the packages using the type `IdOffsetRange{<:AbstractUnitRange}` are not affected by the change in this PR. The only impact this has is to invalidate some precompile statements in `OffsetArrays`, but that's a warning and not a breakage.

jishnub committed:
use dispatch to check eltype

jishnub commented:
I don't understand the test failures, could use some help here

jishnub committed:
move helper functions outside

jishnub committed:
fix definition of IIUR

jishnub committed:
fix getindex for IdOffsetRange

jishnub committed:
Fix IIUR definition

jishnub commented:
Hah tests pass finally!

johnnychen94 commented on base/indices.jl:
```diff
@@ -377,9 +377,22 @@ Represent an AbstractUnitRange `range` as an offset vector such that `range[i] =
 """
-struct IdentityUnitRange{T<:AbstractUnitRange} <: AbstractUnitRange{Int}
-    indices::T
+struct IdentityUnitRange{I<:AbstractUnitRange, T} <: AbstractUnitRange{T}
```
```suggestion
struct IdentityUnitRange{T, I<:AbstractUnitRange} <: AbstractUnitRange{T}
```

For consistency; people would expect the first parameter to be `eltype`. See also `OffsetArrays. IdOffsetRange`

jishnub replied:
Unfortunately this is breaking as some packages appear to rely on the present signature with the first parameter as the range

johnnychen94 replied:
https://github.com/JuliaApproximation/QuasiArrays.jl/blob/2b06122ac40f95344c01c0b5b328b7765ea3f4dc/src/quasiarray.jl#L106

During the GitHub search, I also found an example that already assumes this PR is merged 😆 (with first parameter to be eltype)

cc: @dlfivefifty

dlfivefifty replied:
Ah, guess I never actually hit that path

jishnub replied:
Looking at https://juliahub.com/ui/RepoSearch?q=IdentityUnitRange%7B&r=true, it appears that there are a couple of other packages aside from OffsetArrays and FFTViews. I wonder if this is to be treated as an internal detail that packages should not rely upon?

vtjnash replied:
Should it still at least have `T<:Integer`?

timholy replied:
And add a Julia 2.0 comment about changing the order of the type parameters.

timholy replied:
Another option would be to make
```julia
julia> struct MyRange{I<:AbstractUnitRange{<:Integer}} <: AbstractUnitRange{eltype{I}}
           inds::I
       end
ERROR: TypeError: in Type{...} expression, expected UnionAll, got a value of type typeof(eltype)
Stacktrace:
 [1] top-level scope
   @ REPL[19]:1
```
work.

vtjnash commented on base/indices.jl:
```diff
@@ -377,9 +377,22 @@ Represent an AbstractUnitRange `range` as an offset vector such that `range[i] =
+    indices::I
+
+    function IdentityUnitRange{I,T}(indices::I) where {I<:AbstractUnitRange, T}
+        _eltypecheck(T, eltype(I))
```
```suggestion
    eltype(I) == T || throw(ArgumentError("incorrect eltype for indices of IdentityUnitRange"))
```
